software/grapherrific.livecodescript

```diff
@@ -0,0 +1,1477 @@
+                        put tAkey & ":" & pMain[tDoc][tPtr][tAkey] & cr after pShort[tDoc][tRec]["info"]
+                     end if
+                  end if
+               end repeat
+            end repeat
+         end if
+      end repeat
+   end repeat
+   -- clean up trailing cr
+   repeat for each key tDoc in pShort
+      repeat for each key tRec in pShort[tDoc]
+         repeat for each key tTag in pShort[tDoc][tRec]
+            filter pShort[tDoc][tRec][tTag] without empty
+         end repeat
+      end repeat
+   end repeat
+   -- answer important questions about the state of the graph ------------------------------------------
+   -- undo/redo
+   -- undo is available if a single turn has a single positive pointer
+   -- redo is available if the youngest turn has no positive pointers
+   -- the following counts backwards through the records
+   -- it ignores any record that's not a turn (also ignores undo turns)
+   -- when it gets to the youngest turn it checks for positive pointers
+   -- if no positive pointers in the youngest turn, enable redo & repeat again
+   -- if youngest turn has positive pointers, enable undo & exit (can't redo)
+   -- if redo was enabled, continues counting back to next turn
+   -- continues to record 1
+   -- if no turns have a positive pointer undo stays disabled
+   -- exits repeat as soon as undo enabled
+   repeat for each key tDoc in pMain
+      put "false" into pShort[tDoc][0]["undo"]
+      put "false" into pShort[tDoc][0]["redo"]
+      put the keys of pMain[tDoc] into tRecords
+      sort tRecords numeric descending
+      repeat for each line tRec in tRecords
+         if "turn" is among the keys of pMain[tDoc][tRec] then
+            if pMain[tDoc][tRec]["act"] contains "do" then next repeat
+            if clearNegPairs(pMain[tDoc][tRec]["turn"]) is empty then 
+               --               set the disabled of button "redoButt" stack "primary" to false
+               put "true" into pShort[tDoc][0]["redo"]
+            else
+               --               set the disabled of button "undoButt" stack "primary" to false
+               put "true" into pShort[tDoc][0]["undo"]
+               exit repeat
+            end if
+         end if
+      end repeat
+   end repeat 
+   -- metadata
+   repeat for each key tDoc in pMain
+      put pMain[tDoc][0]["pointerTags"] into pShort[tDoc][0]["pointerTags"]
+   end repeat
+end adjList
+
+command topoSort @pShort,@pTopo,@pNoViz,@pNotes
+   put "command topoSort" into xX["write"];    devLog xX
+   put empty into pTopo -- will contain the sorted nodes
+   put empty into tPlanA -- scope nodes outside of flow
+   put empty into tPlanB -- unscoped nodes at beginning of flow
+   put empty into pNotes -- treats nodes linked in by bond as "footnotes"
+   ---------------- list of seed nodes ----------------------
+   repeat for each key tDoc in pShort
+      put the keys of pShort[tDoc] into pNoViz[tDoc]
+      replace cr with "," in pNoViz[tDoc]
```

Comment:
@matthewmaier Changing the delimiter here seems a bit risky on a referenced parameter as you will need to remember that you did that in your calling handler. I also wonder if `kahnSort` couldn't just deal with lines instead.